Subject:       Changes to boot from plain dm-crypt device
Patch Author:  Orson Teodoro <orsonteodoro@hotmail.com>
Patch Status:  Complete on 3.5.x.x
Date:          2019-01-28 05:01:05.478991171 -0800

diff -urp genkernel.bak/defaults/initrd.scripts genkernel/defaults/initrd.scripts
--- genkernel.bak/defaults/initrd.scripts	2019-01-27 15:56:56.459872616 -0800
+++ genkernel/defaults/initrd.scripts	2019-01-28 05:01:05.478991171 -0800
@@ -1195,7 +1195,7 @@ startVolumes() {
 			then
 				good_msg "ZFS pool ${ZFS_POOL} already imported."
 
-				if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_SWAP}" ]
+				if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_SWAP}" -o -n "${CRYPT_ROOT_PLAIN}" ]
 				then
 					good_msg "LUKS detected. Reimporting ${ZFS_POOL}"
 					/sbin/zpool export -f "${ZFS_POOL}"
@@ -1319,8 +1319,13 @@ openLUKS() {
 			LUKS_DEVICE=$(find_real_device "${LUKS_DEVICE}")
 
 			setup_md_device ${LUKS_DEVICE}
-			cryptsetup isLuks ${LUKS_DEVICE}
-			if [ $? -ne 0 ]
+			if [ -n "${CRYPT_ROOT}" ]
+			then
+				cryptsetup isLuks ${LUKS_DEVICE}
+			else
+				true
+			fi
+			if [ -z "${CRYPT_ROOT_PLAIN}" -a $? -ne 0 ]
 			then
 				bad_msg "The LUKS device ${LUKS_DEVICE} does not contain a LUKS header" ${CRYPT_SILENT}
 				DEV_ERROR=1
@@ -1405,7 +1410,12 @@ openLUKS() {
 					fi
 				fi
 				# At this point, keyfile or not, we're ready!
-				crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				if [ -n "${CRYPT_ROOT_PLAIN}" ]
+				then
+					crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} plainOpen ${CRYPT_ROOT_PLAIN} ${LUKS_NAME}"
+				else
+					crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				fi
 				crypt_filter_ret=$?
 
 				touch ${flag_opened}
@@ -1443,7 +1453,7 @@ startLUKS() {
 	[ -n "${CRYPT_ROOT_KEY}" ] && [ -z "${CRYPT_ROOT_KEYDEV}" ] \
 		&& sleep 6 && bootstrapKey "ROOT"
 
-	if [ -n "${CRYPT_ROOT}" ]; then
+	if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_ROOT_PLAIN}" ]; then
 		openLUKS "root"
 		if [ -n "${REAL_ROOT}" ]
 		then
diff -urp genkernel.bak/defaults/linuxrc genkernel/defaults/linuxrc
--- genkernel.bak/defaults/linuxrc	2019-01-27 15:56:56.458872616 -0800
+++ genkernel/defaults/linuxrc	2019-01-28 03:16:46.612018638 -0800
@@ -216,6 +216,9 @@ do
 		crypt_root=*)
 			CRYPT_ROOT=${x#*=}
 		;;
+		crypt_root_plain=*)
+			CRYPT_ROOT_PLAIN=${x#*=}
+		;;
 		crypt_swap=*)
 			CRYPT_SWAP=${x#*=}
 		;;
